feat(docs): rewrite using-expo-client (#18978)
expo/expo: docs/pages/bare/using-expo-client.md

---
title: Custom native code in Expo Go
sidebar_title: Using Expo Go
---

import { Terminal } from '~/ui/components/Snippet';
import { BoxLink } from '~/ui/components/BoxLink';
import { InlineCode } from '~/components/base/code';
import SnackInline from '~/components/plugins/SnackInline'

> The [Expo Go][expo-go] app is capable of running most React Native apps regardless of how they were bootstrapped. Developing any React Native app in Expo Go is useful because you can distribute your project instantly with the members of your [organization](/accounts/account-types/#organizations), no native build or provisioning required.

Reviewer: ```suggestion
The [Expo Go][expo-go] app is capable of running most React Native apps regardless of how they are bootstrapped. Developing any React Native app in Expo Go is useful because you can distribute your project instantly to the members of your [organization](/accounts/account-types/#organizations), and no native build or provisioning is required.
```

> The main drawback of developing in Expo Go is that it is impossible to send custom native code over-the-air to the Expo Go app, meaning you will need to do one of two things:

Reviewer: ```suggestion
The main drawback of developing in Expo Go is that it is impossible to send custom native code over-the-air to the Expo Go app. This means you will need to do one of the following:
```

> 1. Create a [development build](/development/introduction) to use custom native code, and distribute with [internal distribution](/build/internal-distribution) or TestFlight.

Reviewer: ```suggestion
1. Create a [development build](/development/introduction) to use custom native code, and distribute it with [internal distribution](/build/internal-distribution) or TestFlight.
```

> 2. Conditionally disable unsupported native features and distribute with Expo Go. This guide will demonstrate how achieve this.

Reviewer: ```suggestion
2. Conditionally disable unsupported native features and distribute with Expo Go.

This guide will demonstrate achieving the second option by compatible library versions, detecting whether the code is running in an Expo Go app at run time, native module detection, and so on.
```

I think we should give a proper introduction to what a user can expect from this guide.

## Usage

Inside any React Native app, start a [development server with Expo CLI](/workflow/expo-cli#develop):

> <Terminal cmd={[
> "$ npx expo start",
> ]} cmdCopy="npx expo start" />

Reviewer: ```suggestion
<Terminal cmd={["$ npx expo start"]} />
```

I think this could be a single line since it does not make a difference when represented in the docs.

<BoxLink title={<>Don't have <InlineCode>npx expo start</InlineCode>?</>} href="/bare/installing-expo-modules" description={<>Install and configure the <InlineCode>expo</InlineCode> package in your project.</>} />

> Now you can launch your app in Expo Go by pressing `i` or `a` in the Terminal UI. Some features may cause your app to throw errors because certain native code is missing, continue reading to learn how you can conditionally skip unsupported APIs.

Reviewer: ```suggestion
Then, launch your app in Expo Go by pressing <kbd>i</kbd> or <kbd>a</kbd> in the Terminal UI. Some features may cause your app to throw errors because certain native code is missing, continue reading to learn how you can conditionally skip unsupported APIs.
```

> Unlike `npx react-native start` the command `npx expo start` hosts an app manifest that dev clients like Expo Go can use to load arbitrary projects. Think of an app manifest like the `<head />` element of an `index.html` but for React Native apps. To view this manifest, visit the dev server URL in your web browser.

## Installing libraries

> You'll want to ensure your app is using the most compatible library versions for the project's `react-native` version, this means using `npx expo install` instead of `npm install`. Learn more in [`npx expo install`](/workflow/expo-cli#install).

Reviewer: ```suggestion
Ensure your app uses the most compatible library versions for the project's `react-native` version. This means that use `npx expo install` instead of `npm install` to install libraries. Read more about [`npx expo install`](/workflow/expo-cli#install).
```

## Runtime detection

The easiest way to detect where the JavaScript bundle is running is to check the [`Constants.executionEnvironment`](/versions/latest/sdk/constants/#nativeconstants--properties).

> <Terminal cmd={[
> "$ npx expo install expo-constants",
> ]} cmdCopy="npx expo install expo-constants" />

Reviewer: ```suggestion
<Terminal cmd={[ "$ npx expo install expo-constants"]} />
```

```tsx
import Constants, { ExecutionEnvironment } from 'expo-constants';

// `true` when running in Expo Go.
const isExpoGo = Constants.executionEnvironment === ExecutionEnvironment.StoreClient;
```

You can use this boolean to conditionally require custom native code. Here's an example using the library `react-native-blurhash` which is not available in the Expo Go app:

<SnackInline dependencies={['expo-constants', 'react-native-blurhash']}>

```js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import Constants, { ExecutionEnvironment } from 'expo-constants';

// `true` when running in Expo Go.
const isExpoGo = Constants.executionEnvironment === ExecutionEnvironment.StoreClient;

let Blurhash;

> 

Reviewer: ```suggestion
// Create a fallback for Expo Go
```

Can we add a comment here about creating a fallback for Expo Go? I think it will be a nice addition.

        },
        props.style,
      ]}>
      <Text>(Blurhash not available)</Text>
    </View>
  );
} else {
  // Conditionally require this module to prevent Metro from throwing warnings.
  Blurhash = require('react-native-blurhash').Blurhash;
}

export default function App() {
  return <Blurhash blurhash="LGFFaXYk^6#M@-5c,1J5@[or[Q6." style={{ flex: 1 }} />;
}
```

</SnackInline>

> If you run this code in the Expo Go app you'll see the fallback view, if you were to build this locally with the [Expo CLI run commands](/workflow/expo-cli#compiling) then you'd see the native blur hash view.

Reviewer: ```suggestion
If you run this code in the Expo Go app, you'll see the fallback view. If you are able to build this locally with the [Expo CLI run commands](/workflow/expo-cli#compiling) then you'll see the native blur hash view.
```

Let's use as much present tense as we can as per our writing style guide.

## Native module detection

> Native modules are added to the JavaScript global object by the runtime, this means you can conditionally check if they exist to ensure functionality:

Reviewer: ```suggestion
Native modules are added to the JavaScript global object at the runtime. This means you can conditionally check if they exist to ensure functionality:
```

```js
import { NativeModules } from 'react-native';

const isAvailable = !!NativeModules.MyAnalytics;
```

> This is nice because it ensures the native module _must_ be installed and linked. Two issues with this solution:

Reviewer: ```suggestion
The above code snippet ensures the native module _must_ be installed and linked. However, there are two issues with this solution:
```

1. You need to know the native module name ahead of time.

> 2. You likely want an error to be thrown when a native module is missing in a custom build, this helps you determine if there was a native linking issue.

Reviewer: ```suggestion
2. You likely want an error to be thrown when a native module is missing in a custom build. This helps you determine if there is a native linking issue.
```

Reviewer: Not a suggestion but a comment. This sounds like too technical (which is not a bad thing). So can we give an example or include an example on how to determine if there is a native linking issue.

## Optional imports

> [Metro bundler](/guides/customizing-metro) has built-in support for optional imports. Optional imports refer to wrapping a `require` statement with a `try/catch` to prevent an error from being thrown when the requested module is missing:

Reviewer: ```suggestion
Optional imports are supported by [Metro bundler](/guides/customizing-metro). They refer to wrapping a `require` statement with a `try/catch` to prevent an error from being thrown when the requested module is missing:
```

<SnackInline dependencies={['expo-constants', 'react-native-blurhash']}>

```js
import React from 'react';
import { View } from 'react-native';

let Blurhash;

try {
  Blurhash = require('react-native-blurhash').Blurhash;
} catch {
  Blurhash = View;
}

export default function App() {
  return <Blurhash blurhash="LGFFaXYk^6#M@-5c,1J5@[or[Q6." style={{ flex: 1 }} />;
}
```

</SnackInline>

> This method is the least reliable because there are a number of reasons that a `require` statement might throw an error. There could be an internal error, the module could be missing, the native module could be linked incorrectly, etc. Ultimately you should try to avoid using this method.

Reviewer: ```suggestion
This method is the least reliable because there are several reasons that a `require` statement might throw an error. For example, there could be an internal error, the module could be missing, the native module could be linked incorrectly, and so on. You should avoid using this method.
```

Let's be authoritative here with the tone for avoiding this method.

## **Deprecated**: use the `.expo.[js/json/ts/tsx]` extension to provide Expo Go specific fallbacks

> The `.expo` extension is removed in SDK 41. Consider using [conditional inline requires](#use-conditional-inline-requires-to-provide-fallbacks) instead, and read [expo.fyi/expo-extension-migration](https://expo.fyi/expo-extension-migration) for specific guidance on migrating away.

Reviewer: ```suggestion
The `.expo` extension is removed in SDK 41. Instead, consider using [conditional inline requires](#use-conditional-inline-requires-to-provide-fallbacks), and read [expo.fyi/expo-extension-migration](https://expo.fyi/expo-extension-migration) for specific guidance on migration.
```

Unrelated to PR, but I think this is a good opportunity to improve the verbiage here too.
